test/Microsoft.ML.Functional.Tests/DataIO.cs

```diff
@@ -0,0 +1,353 @@
+        {
+            var filePath = GetOutputPath(Path.GetRandomFileName());
+            using (var file = File.Create(filePath))
+                mlContext.Data.SaveAsBinary(data, file);
+
+            return filePath;
+        }
+        #endregion
+
+        #region ToyDataset
+        private void ToyDatasetsAreEqual(MLContext mlContext, IDataView data1, IDataView data2)
+        {
+            // Validate that the two Schemas are the same
+            Common.AssertEqual(data1.Schema, data2.Schema);
+
+            // Define how to serialize the IDataView to objects
+            var enumerable1 = mlContext.CreateEnumerable<ToyDataset>(data1, true);
+            var enumerable2 = mlContext.CreateEnumerable<ToyDataset>(data2, true);
+
+            AssertEqual(enumerable1, enumerable2);
+        }
+
+        private void AssertEqual(IEnumerable<ToyDataset> data1, IEnumerable<ToyDataset> data2)
+        {
+            Assert.NotNull(data1);
+            Assert.NotNull(data2);
+            Assert.Equal(data1.Count(), data2.Count());
+
+            foreach (var rowPair in data1.Zip(data2, Tuple.Create))
+            {
+                AssertEqual(rowPair.Item1, rowPair.Item2);
+            }
+        }
+
+            private void ValidateToyDataset(IDataView toyDataset)
+        {
+            var toyClassProperties = typeof(ToyDataset).GetProperties();
+
+            // Check that the schema is of the right size
+            Assert.Equal(17, toyDataset.Schema.Count);
+
+            // Create a lookup table for the types and counts of all properties
+            var types = new Dictionary<string, Type>();
+            var counts = new Dictionary<string, int>();
+            foreach (var property in toyClassProperties)
+            {
+                if (!property.PropertyType.IsArray)
+                    types[property.Name] = property.PropertyType;
+                else
+                {
+                    // Construct a VBuffer type for the array
+                    var vBufferType = typeof(VBuffer<>);
+                    Type[] typeArgs = { property.PropertyType.GetElementType() };
+                    Activator.CreateInstance(property.PropertyType.GetElementType());
+                    types[property.Name] = vBufferType.MakeGenericType(typeArgs);
+                }
+                    
+                counts[property.Name] = 0;
+            }
+
+            foreach (var column in toyDataset.Schema)
+            {
+                Assert.True(types.ContainsKey(column.Name));
+                Assert.Equal(1, ++counts[column.Name]);
+                Assert.Equal(types[column.Name], column.Type.RawType);
+            }
+
+            // Make sure we didn't miss any columns
+            foreach (var value in counts.Values)
+                Assert.Equal(1, value);
+        }
+
+        private IEnumerable<ToyDataset> GenerateToyDataset(int numExamples = 5, int seed = 1)
+        {
+            var rng = new Random(seed);
```

Comment:
>ng = new Random(seed); [](start = 17, length = 22)

we can't get random from context?

Reply:
Only from `IHostEnvironment`, I think. That isn't accessible outside the library.

---
In reply to: [256211132](https://github.com/dotnet/machinelearning/pull/2518#discussion_r256211132) [](ancestors = 256211132)